File: app/api/budgets.py
"""
Budget Management API endpoints
"""
from ninja import Router
from typing import List, Optional
from pydantic import BaseModel
from datetime import date
from ..models import Budget, Category
from ..services.budget_service import budget_service

router = Router(tags=["budgets"])


class BudgetIn(BaseModel):
    category_id: int
    amount: str
    period: str = "monthly"
    start_date: date
    end_date: date


class BudgetOut(BaseModel):
    id: int
    category_id: int
    category_name: str
    amount: str
    period: str
    start_date: str
    end_date: str
    is_active: bool
    created_at: str


class BudgetStatusOut(BaseModel):
    budget_id: int
    category: str
    amount: float
    spent: float
    remaining: float
    percentage: float
    status: str
    period: str
    start_date: str
    end_date: str


@router.get("", response=List[BudgetOut], summary="List all budgets")
def list_budgets(request):
    """Get all budgets"""
    budgets = Budget.objects.select_related('category').all()
    return [
        {
            "id": b.id,
            "category_id": b.category.id,
            "category_name": b.category.name,
            "amount": str(b.amount),
            "period": b.period,
            "start_date": b.start_date.isoformat(),
            "end_date": b.end_date.isoformat(),
            "is_active": b.is_active,
            "created_at": b.created_at.isoformat(),
        }
        for b in budgets
    ]


@router.get("/{budget_id}", response=BudgetOut, summary="Get budget by ID")
def get_budget(request, budget_id: int):
    """Get a specific budget"""
    budget = Budget.objects.select_related('category').get(id=budget_id)
    return {
        "id": budget.id,
        "category_id": budget.category.id,
        "category_name": budget.category.name,
        "amount": str(budget.amount),
        "period": budget.period,
        "start_date": budget.start_date.isoformat(),
        "end_date": budget.end_date.isoformat(),
        "is_active": budget.is_active,
        "created_at": budget.created_at.isoformat(),
    }


@router.get("/{budget_id}/status", response=BudgetStatusOut, summary="Get budget status")
def get_budget_status(request, budget_id: int):
    """Get budget status with spent/remaining amounts"""
    budget = Budget.objects.select_related('category').get(id=budget_id)
    status = budget_service.get_budget_status(budget)
    return status


@router.post("", response=BudgetOut, summary="Create new budget")
def create_budget(request, data: BudgetIn):
    """Create a new budget"""
    category = Category.objects.get(id=data.category_id)
    budget = Budget.objects.create(
        category=category,
        amount=data.amount,
        period=data.period,
        start_date=data.start_date,
        end_date=data.end_date,
    )
    return {
        "id": budget.id,
        "category_id": budget.category.id,
        "category_name": budget.category.name,
        "amount": str(budget.amount),
        "period": budget.period,
        "start_date": budget.start_date.isoformat(),
        "end_date": budget.end_date.isoformat(),
        "is_active": budget.is_active,
        "created_at": budget.created_at.isoformat(),
    }


@router.put("/{budget_id}", response=BudgetOut, summary="Update budget")
def update_budget(request, budget_id: int, data: BudgetIn):
    """Update an existing budget"""
    budget = Budget.objects.get(id=budget_id)
    budget.category = Category.objects.get(id=data.category_id)
    budget.amount = data.amount
    budget.period = data.period
    budget.start_date = data.start_date
    budget.end_date = data.end_date
    budget.save()
    
    return {
        "id": budget.id,
        "category_id": budget.category.id,
        "category_name": budget.category.name,
        "amount": str(budget.amount),
        "period": budget.period,
        "start_date": budget.start_date.isoformat(),
        "end_date": budget.end_date.isoformat(),
        "is_active": budget.is_active,
        "created_at": budget.created_at.isoformat(),
    }


@router.delete("/{budget_id}", summary="Delete budget")
def delete_budget(request, budget_id: int):
    """Delete a budget"""
    budget = Budget.objects.get(id=budget_id)
    budget.delete()
    return {"success": True, "message": "Budget deleted"}

File: static/js/budgets.js
// Budget Management JavaScript
document.addEventListener('DOMContentLoaded', () => {
    loadBudgets();
    loadCategories();
    
    // Set default dates
    const today = new Date();
    const startDate = new Date(today.getFullYear(), today.getMonth(), 1);
    const endDate = new Date(today.getFullYear(), today.getMonth() + 1, 0);
    
    document.getElementById('budgetStartDate').value = startDate.toISOString().split('T')[0];
    document.getElementById('budgetEndDate').value = endDate.toISOString().split('T')[0];
    
    // Form submit
    const form = document.getElementById('addBudgetForm');
    if (form) {
        form.addEventListener('submit', handleAddBudget);
    }
});

async function loadCategories() {
    try {
        const categories = await window.API.call('/categories');
        const select = document.getElementById('budgetCategory');
        if (select && categories) {
            const items = Array.isArray(categories) ? categories : (categories.items || []);
            items.forEach(cat => {
                const option = document.createElement('option');
                option.value = cat.id;
                option.textContent = cat.name;
                select.appendChild(option);
            });
        }
    } catch (error) {
        console.error('Error loading categories:', error);
    }
}

async function loadBudgets() {
    const container = document.getElementById('budgetsContainer');
    if (!container) return;
    
    try {
        const budgets = await window.API.call('/budgets');
        if (budgets && budgets.length > 0) {
            displayBudgets(budgets);
        } else {
            container.innerHTML = `
                <div class="col-12">
                    <div class="card border-0 shadow-sm">
                        <div class="card-body">
                            <div class="text-center py-5">
                                <i class="bi bi-inbox fs-1 text-muted d-block mb-3"></i>
                                <p class="text-muted mb-0">Chưa có ngân sách nào. Hãy tạo ngân sách đầu tiên của bạn!</p>
                            </div>
                        </div>
                    </div>
                </div>
            `;
        }
    } catch (error) {
        console.error('Error loading budgets:', error);
        container.innerHTML = `
            <div class="col-12">
                <div class="alert alert-danger" role="alert">
                    <i class="bi bi-exclamation-triangle"></i> Lỗi khi tải danh sách ngân sách.
                </div>
            </div>
        `;
    }
}

function displayBudgets(budgets) {
    const container = document.getElementById('budgetsContainer');
    
    const html = budgets.map(budget => {
        const periodText = {
            'monthly': 'Hàng tháng',
            'weekly': 'Hàng tuần',
            'yearly': 'Hàng năm'
        }[budget.period] || budget.period;
        
        return `
            <div class="col-md-6 col-lg-4">
                <div class="card border-0 shadow-sm h-100">
                    <div class="card-body">
                        <div class="d-flex justify-content-between align-items-start mb-3">
                            <h6 class="card-title mb-0">
                                <span class="badge bg-primary">${budget.category_name}</span>
                            </h6>
                            <span class="badge bg-${budget.is_active ? 'success' : 'secondary'}">
                                ${budget.is_active ? 'Đang hoạt động' : 'Tạm dừng'}
                            </span>
                        </div>
                        <p class="mb-2">
                            <strong>Số tiền:</strong> 
                            <span class="text-primary fs-5">${formatCurrency(budget.amount)}</span>
                        </p>
                        <p class="mb-2">
                            <strong>Chu kỳ:</strong> ${periodText}
                        </p>
                        <p class="mb-2 small text-muted">
                            <i class="bi bi-calendar"></i> 
                            ${formatDate(budget.start_date)} - ${formatDate(budget.end_date)}
                        </p>
                        <div class="mt-3">
                            <button class="btn btn-sm btn-outline-primary" onclick="viewBudgetStatus(${budget.id})">
                                <i class="bi bi-graph-up"></i> Xem trạng thái
                            </button>
                            <button class="btn btn-sm btn-outline-danger" onclick="deleteBudget(${budget.id})">
                                <i class="bi bi-trash"></i> Xóa
                            </button>
                        </div>
                    </div>
                </div>
            </div>
        `;
    }).join('');
    
    container.innerHTML = html;
}

async function handleAddBudget(e) {
    e.preventDefault();
    
    const categoryId = document.getElementById('budgetCategory').value;
    const amount = document.getElementById('budgetAmount').value;
    const period = document.getElementById('budgetPeriod').value;
    const startDate = document.getElementById('budgetStartDate').value;
    const endDate = document.getElementById('budgetEndDate').value;
    
    if (!categoryId || !amount || !startDate || !endDate) {
        showAlert('Vui lòng điền đầy đủ thông tin', 'warning');
        return;
    }
    
    try {
        const result = await window.API.call('/budgets', {
            method: 'POST',
            body: JSON.stringify({
                category_id: parseInt(categoryId),
                amount: amount,
                period: period,
                start_date: startDate,
                end_date: endDate
            })
        });
        
        if (result) {
            showAlert('Tạo ngân sách thành công!', 'success');
            // Close modal
            const modal = bootstrap.Modal.getInstance(document.getElementById('addBudgetModal'));
            if (modal) modal.hide();
            // Reset form
            document.getElementById('addBudgetForm').reset();
            // Reload budgets
            loadBudgets();
        }
    } catch (error) {
        console.error('Error creating budget:', error);
        showAlert('Lỗi khi tạo ngân sách. Vui lòng thử lại.', 'danger');
    }
}

async function viewBudgetStatus(budgetId) {
    try {
        const status = await window.API.call(`/budgets/${budgetId}/status`);
        if (status) {
            const percentage = status.percentage || 0;
            const statusClass = percentage >= 100 ? 'danger' : (percentage >= 80 ? 'warning' : 'success');
            
            const html = `
                <div class="alert alert-${statusClass}" role="alert">
                    <h6><i class="bi bi-graph-up"></i> Trạng thái Ngân sách</h6>
                    <p class="mb-2"><strong>Danh mục:</strong> ${status.category}</p>
                    <p class="mb-2"><strong>Ngân sách:</strong> ${formatCurrency(status.amount)}</p>
                    <p class="mb-2"><strong>Đã chi:</strong> ${formatCurrency(status.spent)}</p>
                    <p class="mb-2"><strong>Còn lại:</strong> ${formatCurrency(status.remaining)}</p>
                    <div class="progress mt-2" style="height: 25px;">
                        <div class="progress-bar bg-${statusClass}" role="progressbar" style="width: ${Math.min(percentage, 100)}%">
                            ${percentage.toFixed(1)}%
                        </div>
                    </div>
                </div>
            `;
            
            showModal('Trạng thái Ngân sách', html);
        }
    } catch (error) {
        console.error('Error loading budget status:', error);
        showAlert('Lỗi khi tải trạng thái ngân sách', 'danger');
    }
}

async function deleteBudget(budgetId) {
    if (!confirm('Bạn có chắc chắn muốn xóa ngân sách này?')) {
        return;
    }
    
    try {
        const result = await window.API.call(`/budgets/${budgetId}`, {
            method: 'DELETE'
        });
        
        if (result && result.success) {
            showAlert('Xóa ngân sách thành công!', 'success');
            loadBudgets();
        }
    } catch (error) {
        console.error('Error deleting budget:', error);
        showAlert('Lỗi khi xóa ngân sách', 'danger');
    }
}

function showAlert(message, type = 'info') {
    const alertDiv = document.createElement('div');
    alertDiv.className = `alert alert-${type} alert-dismissible fade show position-fixed top-0 start-50 translate-middle-x mt-3`;
    alertDiv.style.zIndex = '9999';
    alertDiv.innerHTML = `
        ${message}
        <button type="button" class="btn-close" data-bs-dismiss="alert"></button>
    `;
    document.body.appendChild(alertDiv);
    
    setTimeout(() => {
        alertDiv.remove();
    }, 3000);
}

function showModal(title, content) {
    // Create and show modal
    const modalHtml = `
        <div class="modal fade" id="infoModal" tabindex="-1">
            <div class="modal-dialog">
                <div class="modal-content">
                    <div class="modal-header">
                        <h5 class="modal-title">${title}</h5>
                        <button type="button" class="btn-close" data-bs-dismiss="modal"></button>
                    </div>
                    <div class="modal-body">
                        ${content}
                    </div>
                    <div class="modal-footer">
                        <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Đóng</button>
                    </div>
                </div>
            </div>
        </div>
    `;
    
    // Remove existing modal if any
    const existing = document.getElementById('infoModal');
    if (existing) existing.remove();
    
    document.body.insertAdjacentHTML('beforeend', modalHtml);
    const modal = new bootstrap.Modal(document.getElementById('infoModal'));
    modal.show();
}

function formatCurrency(amount) {
    return new Intl.NumberFormat('vi-VN', {
        style: 'currency',
        currency: 'VND'
    }).format(parseFloat(amount || 0));
}

function formatDate(dateString) {
    if (!dateString) return 'N/A';
    const date = new Date(dateString);
    return date.toLocaleDateString('vi-VN');
}

